source/menu/menu_coop.qc

```diff
@@ -0,0 +1,299 @@
+
+float player_name_cursor;
+string player_name;
+
+float server_ip_cursor;
+string server_ip;
+
+float server_hostname_cursor;
+string server_hostname;
+
+float server_password_cursor;
+string server_password;
+
+float server_port_cursor;
+string server_port;
+
+int num_cached_servers;
+
+void() Menu_Coop_Name_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		player_name = cvar_string("name");
+		player_name_cursor = strlen(player_name);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("name", player_name);
+	}
+};
+
+void() Menu_Coop_Hostname_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		server_hostname = cvar_string("hostname");
+		server_hostname_cursor = strlen(server_hostname);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("hostname", server_hostname);
+	}
+};
+
+void() Menu_Coop_Port_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		server_port = cvar_string("sv_port");
+		server_port_cursor = strlen(server_port);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("sv_port", server_port);
+	}
+};
+
+void() Menu_Coop_Password_Update =
+{
+	static float initialized;
+	if (!initialized)
+	{
+		server_password = cvar_string("password");
+		server_password_cursor = strlen(server_password);
+		initialized = true;
+	}
+	else
+	{
+		cvar_set("password", server_password);
+	}
+};
+
+// this flushes the server cache
+void() Menu_Coop_Browse_Refresh =
+{
+	static float refreshtime;
+
+	if (refreshtime > time)
+		return;
+
+	refreshtime = time + 0.5;
+
+	resethostcachemasks();
+	sethostcachemaskstring(0, gethostcacheindexforkey("gamedir"), cvar_string("game"), SLIST_TEST_EQUAL);
+	sethostcachesort(gethostcacheindexforkey("ping"), false);
+	refreshhostcache(true);
+	resorthostcache();
+
+	num_cached_servers = (int)gethostcachevalue(SLIST_HOSTCACHEVIEWCOUNT);
+};
+
+// this updates the current server cache
+void() Menu_Coop_Browse_Update =
+{
+	static float refreshtime;
+	static float initialized;
+
+	if (!initialized)
+	{
+		Menu_Coop_Browse_Refresh();
+		initialized = true;
+	}
+
+	if (refreshtime > time)
+		return;
+
+	refreshtime = time + 0.5;
+
+	num_cached_servers = (int)gethostcachevalue(SLIST_HOSTCACHEVIEWCOUNT);
+};
+
+void() Menu_Coop_Browse =
+{
+	Menu_Coop_Browse_Update();
+
+	Menu_DrawBackground();
+
+	Menu_DrawTitle("BROWSE SERVERS");
+
+	// draw server list
+	sui_set_align([SUI_ALIGN_CENTER, SUI_ALIGN_CENTER]);
+	static vector serverlist_scrollofs;
+	vector vsize = sui_current_frame_size();
+	Menu_ServerList("cbm_serverlist", [0, 0], [vsize.x, vsize.y / 2], serverlist_scrollofs, num_cached_servers);
+
```

Comment:
We should have some kind of text on this menu indicating the server browser doesn't support non-mouse input.

Reply:
Addressed in [5c51661](https://github.com/nzp-team/quakec/pull/90/commits/5c516618ff54cc82ab2dc9de53afc151d5968052)